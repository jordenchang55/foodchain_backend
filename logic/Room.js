/* eslint-disable no-underscore-dangle */

import Game from './Game';

export class Room {
    constructor(eventManager, capacity = 5) {
        this.eventManager = eventManager;
        this.capacity = capacity;
        this.playerStatus = {};
        this.spectaculars = {};
        this.isPlaying = false;

        this.setupEventListeners();
    }

    addPlayer(username) {
        const playerNum = Object.keys(this.playerStatus).length;
        if (playerNum >= this.capacity || this.isPlaying) {
            this.spectaculars[username] = {};
        } else {
            const seat = Object.values(this.playerStatus)
                .map((p) => p.index)
                .reduce((prev, curr) => (curr === prev ? (curr + 1) : Math.min(prev, curr)), 0);
            this.playerStatus[username] = {
                prepared: false,
                index: seat,
            };
        }
        this.notifyPlayerList();
    }

    removePlayer(username) {
        if (this.playerStatus[username]) {
            delete this.playerStatus[username];
        } else {
            delete this.spectaculars[username];
        }
        this.notifyPlayerList();
    }

    setupEventListeners() {
        this.eventManager.on('$$player_add', (username) => {
            this.addPlayer(username);
        });
        this.eventManager.on('$$player_remove', (username) => {
            this.removePlayer(username);
        });
        this.eventManager.on(
            'preparation',
            (u, msg) => this.onPlayerPrepared(u, msg.prepared),
        );
        this.eventManager.on(
            'exchange_place',
            (u, msg) => this.onPlayerExchanged(u, msg.index),
        );
    }

    onPlayerPrepared(username, prepared) {
        if (this.playerStatus[username]) {
            this.playerStatus[username].prepared = prepared;
        }
        this.notifyPlayerList();
        if (Object.values(this.playerStatus).length >= 2 && Object.values(this.playerStatus).every((p) => p.prepared)) {
            this.launchGame();
        }
    }

    onPlayerExchanged(username, index) {
        if (this.playerStatus[username]) {
            this.playerStatus[username].index = index;
        }
        this.notifyPlayerList();
    }

    notifyPlayerList() {
        this.eventManager.notifyAll('player_list_update', {
            players: this.playerStatus,
            spectaculars: Object.keys(this.spectaculars),
        });
    }

    launchGame() {
        this.isPlaying = true;
        this.eventManager.notifyAll('launch', {
            players: this.playerStatus,
            spectaculars: Object.keys(this.spectaculars),
        });
        this.game = new Game(this.eventManager);
        this.game.initialize(Object.keys(this.playerStatus));
        this.game.setup();
    }
}
