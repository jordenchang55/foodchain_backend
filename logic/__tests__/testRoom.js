import EventManager from '../../models/EventManager';
import { Room } from '../Room';

jest.mock('../../models/EventManager');

describe('test Room', () => {
    let eventMgr;
    beforeEach(() => {
        jest.spyOn(global.Math, 'random').mockReturnValue(0.5);
        EventManager.mockClear();
        eventMgr = new EventManager();
    });
    describe('add player', () => {
        it('success', () => {
            const room = new Room(eventMgr);
            expect(EventManager).toHaveBeenCalledTimes(1);

            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            expect(mockEventMgr.on).toBeCalledWith('preparation', expect.any(Function));
            expect(mockEventMgr.notifyAll).toBeCalledTimes(1);
            expect(mockEventMgr.notifyAll).toBeCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                },
                spectaculars: [],
            });
        });
        it('multiple', () => {
            const room = new Room(eventMgr);
            expect(EventManager).toHaveBeenCalledTimes(1);

            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            expect(mockEventMgr.notifyAll).toBeCalledTimes(1);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                },
                spectaculars: [],
            });
            room.addPlayer('user2');
            expect(mockEventMgr.notifyAll).toBeCalledTimes(2);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                    user2: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
        });
        it('add vacant seat', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.addPlayer('user3');
            room.addPlayer('user4');
            room.removePlayer('user2');
            room.removePlayer('user3');
            room.addPlayer('new_user');
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                    user4: {
                        prepared: false,
                        index: 3,
                    },
                    new_user: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
        });
        it('add vacant seat at first', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.addPlayer('user3');
            room.addPlayer('user4');
            room.removePlayer('user1');
            room.removePlayer('user3');
            room.addPlayer('new_user');
            expect(mockEventMgr.notifyAll.mock.calls).toMatchSnapshot();
        });
        it('player full', () => {
            const room = new Room(eventMgr, 2);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.addPlayer('user3');
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                    user2: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: ['user3'],
            });
        });
        it('add player after playing', () => {
            const room = new Room(eventMgr, 3);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.onPlayerPrepared('user1', true);
            room.onPlayerPrepared('user2', true);
            expect(mockEventMgr.notifyAll.mock.calls).toMatchSnapshot();
            room.addPlayer('user3');
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: true,
                        index: 0,
                    },
                    user2: {
                        prepared: true,
                        index: 1,
                    },
                },
                spectaculars: ['user3'],
            });
        });
    });
    describe('remove player', () => {
        it('existing player', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.removePlayer('user1');
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user2: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
        });
        it('not existing player', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.removePlayer('nobody');
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                    user2: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
        });
    });
    describe('test preparation change', () => {
        it('set prepared', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.onPlayerPrepared('user1', true);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: true,
                        index: 0,
                    },
                },
                spectaculars: [],
            });
        });
        it('set prepared multiple user', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.onPlayerPrepared('user2', true);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                    user2: {
                        prepared: true,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
        });
        it('cancel prepared', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.onPlayerPrepared('user1', true);
            room.onPlayerPrepared('user1', false);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 0,
                    },
                },
                spectaculars: [],
            });
        });
        it('all prepared multiple players', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.addPlayer('user2');
            room.onPlayerPrepared('user1', true);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: true,
                        index: 0,
                    },
                    user2: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
            room.onPlayerPrepared('user2', true);
            expect(mockEventMgr.notifyAll.mock.calls).toMatchSnapshot();
        });
    });
    describe('player exchange restaurant', () => {
        it('exchange from 0 t0 1', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.onPlayerExchanged('user1', 1);
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 1,
                    },
                },
                spectaculars: [],
            });
        });
        it('new player joined after exchange', () => {
            const room = new Room(eventMgr);
            const mockEventMgr = EventManager.mock.instances[0];
            room.addPlayer('user1');
            room.onPlayerExchanged('user1', 2);
            room.addPlayer('user2');
            expect(mockEventMgr.notifyAll).toHaveBeenLastCalledWith('player_list_update', {
                players: {
                    user1: {
                        prepared: false,
                        index: 2,
                    },
                    user2: {
                        prepared: false,
                        index: 0,
                    },
                },
                spectaculars: [],
            });
        });
    });
});
