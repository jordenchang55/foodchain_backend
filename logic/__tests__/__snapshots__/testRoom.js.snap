// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Room add player add player after playing 1`] = `
Array [
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "launch",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "setup_map",
    Object {
      "tiles": Array [
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 0,
            "yTile": 0,
          },
          "tileId": "M01",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 1,
            "yTile": 0,
          },
          "tileId": "M18",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 2,
            "yTile": 0,
          },
          "tileId": "M02",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 0,
            "yTile": 1,
          },
          "tileId": "M14",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 1,
            "yTile": 1,
          },
          "tileId": "M03",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 2,
            "yTile": 1,
          },
          "tileId": "M12",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 0,
            "yTile": 2,
          },
          "tileId": "M04",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 1,
            "yTile": 2,
          },
          "tileId": "M16",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 2,
            "yTile": 2,
          },
          "tileId": "M05",
        },
      ],
    },
  ],
  Array [
    "pool_update",
    Object {
      "ads": Array [
        "Mk01",
        "Mk02",
        "Mk03",
        "Mk04",
        "Mk05",
        "Mk06",
        "Mk07",
        "Mk08",
        "Mk09",
        "Mk10",
        "Mk11",
        "Mk13",
        "Mk14",
      ],
      "employees": Object {
        "E01": 12,
        "E02": 18,
        "E03": 12,
        "E04": 12,
        "E05": 12,
        "E06": 12,
        "E07": 12,
        "E08": 12,
        "E09": 6,
        "E10": 12,
        "E11": 1,
        "E12": 6,
        "E13": 6,
        "E14": 6,
        "E15": 6,
        "E16": 6,
        "E17": 6,
        "E18": 6,
        "E19": 6,
        "E20": 6,
        "E21": 6,
        "E22": 6,
        "E23": 1,
        "E24": 1,
        "E25": 1,
        "E26": 6,
        "E27": 3,
        "E28": 1,
        "E29": 1,
        "E30": 1,
        "E31": 1,
        "E32": 1,
      },
      "gardenNumber": 8,
      "houses": Array [
        "H01",
        "H03",
        "H06",
        "H09",
        "H11",
        "H14",
        "H17",
        "H19",
      ],
      "milestones": Object {
        "achieved": Object {},
        "new": Object {},
        "remain": Array [
          "Mi01",
          "Mi02",
          "Mi03",
          "Mi04",
          "Mi05",
          "Mi06",
          "Mi07",
          "Mi08",
          "Mi09",
          "Mi10",
          "Mi11",
          "Mi12",
          "Mi13",
          "Mi14",
          "Mi15",
          "Mi16",
          "Mi17",
          "Mi18",
        ],
      },
      "money": 100,
    },
  ],
  Array [
    "order_decision",
    Object {
      "available": Array [],
      "selected": Array [
        "user1",
        "user2",
      ],
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user2",
      "skippable": true,
    },
  ],
]
`;

exports[`test Room add player add vacant seat at first 1`] = `
Array [
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "new_user": Object {
          "index": 0,
          "prepared": false,
        },
        "user2": Object {
          "index": 1,
          "prepared": false,
        },
        "user4": Object {
          "index": 3,
          "prepared": false,
        },
      },
      "spectaculars": Array [],
    },
  ],
]
`;

exports[`test Room test preparation change all prepared multiple players 1`] = `
Array [
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "player_list_update",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "launch",
    Object {
      "players": Object {
        "user1": Object {
          "index": 0,
          "prepared": true,
        },
        "user2": Object {
          "index": 1,
          "prepared": true,
        },
      },
      "spectaculars": Array [],
    },
  ],
  Array [
    "setup_map",
    Object {
      "tiles": Array [
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 0,
            "yTile": 0,
          },
          "tileId": "M01",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 1,
            "yTile": 0,
          },
          "tileId": "M18",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 2,
            "yTile": 0,
          },
          "tileId": "M02",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 0,
            "yTile": 1,
          },
          "tileId": "M14",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 1,
            "yTile": 1,
          },
          "tileId": "M03",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 2,
            "yTile": 1,
          },
          "tileId": "M12",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 0,
            "yTile": 2,
          },
          "tileId": "M04",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 1,
            "yTile": 2,
          },
          "tileId": "M16",
        },
        Object {
          "direction": 2,
          "position": Object {
            "xTile": 2,
            "yTile": 2,
          },
          "tileId": "M05",
        },
      ],
    },
  ],
  Array [
    "pool_update",
    Object {
      "ads": Array [
        "Mk01",
        "Mk02",
        "Mk03",
        "Mk04",
        "Mk05",
        "Mk06",
        "Mk07",
        "Mk08",
        "Mk09",
        "Mk10",
        "Mk11",
        "Mk13",
        "Mk14",
      ],
      "employees": Object {
        "E01": 12,
        "E02": 18,
        "E03": 12,
        "E04": 12,
        "E05": 12,
        "E06": 12,
        "E07": 12,
        "E08": 12,
        "E09": 6,
        "E10": 12,
        "E11": 1,
        "E12": 6,
        "E13": 6,
        "E14": 6,
        "E15": 6,
        "E16": 6,
        "E17": 6,
        "E18": 6,
        "E19": 6,
        "E20": 6,
        "E21": 6,
        "E22": 6,
        "E23": 1,
        "E24": 1,
        "E25": 1,
        "E26": 6,
        "E27": 3,
        "E28": 1,
        "E29": 1,
        "E30": 1,
        "E31": 1,
        "E32": 1,
      },
      "gardenNumber": 8,
      "houses": Array [
        "H01",
        "H03",
        "H06",
        "H09",
        "H11",
        "H14",
        "H17",
        "H19",
      ],
      "milestones": Object {
        "achieved": Object {},
        "new": Object {},
        "remain": Array [
          "Mi01",
          "Mi02",
          "Mi03",
          "Mi04",
          "Mi05",
          "Mi06",
          "Mi07",
          "Mi08",
          "Mi09",
          "Mi10",
          "Mi11",
          "Mi12",
          "Mi13",
          "Mi14",
          "Mi15",
          "Mi16",
          "Mi17",
          "Mi18",
        ],
      },
      "money": 100,
    },
  ],
  Array [
    "order_decision",
    Object {
      "available": Array [],
      "selected": Array [
        "user1",
        "user2",
      ],
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user2",
      "skippable": true,
    },
  ],
]
`;
